core/base/v7/inc/ROOT/TDirectoryEntry.hxx

```diff
@@ -75,7 +70,13 @@ public:
 
 template<class U>
 std::shared_ptr<U> TDirectoryEntry::CastPointer() const {
-  if (auto ptr = fType->DynamicCast(TClass::GetClass(typeid(U)), fObj.get()))
+  using cast_t = bool(const std::type_info*, const std::type_info*, void**, unsigned);
+
+  void* ptr = fObj.get();
+  void** vtbl = reinterpret_cast<void** const*>(&typeid(U))[0];
+  cast_t* cast = reinterpret_cast<cast_t**>(vtbl)[4];
+
+  if (cast(&typeid(U), fType, &ptr, 0))
```

Comment:
> Since this is likely to be platform dependent

Ofc.

>  does it work as-is on Windows Microsoft compiler

No doubt about that it doesn't.

>  can we re-add the previous code for the platform not covered?

We can but... later.